File: src/main/kotlin/io/titandata/titan/providers/generic/Commit.kt
/*
 * Copyright (c) 2019 by Delphix. All rights reserved.
 */

package io.titandata.titan.providers.generic

import io.titandata.client.apis.CommitsApi
import io.titandata.client.apis.RepositoriesApi
import io.titandata.models.Commit
import io.titandata.titan.providers.Metadata
import io.titandata.titan.providers.Metadata.Companion.toMap
import java.util.*

class Commit (
    private val user: String,
    private val email: String,
    private val repositoriesApi: RepositoriesApi = RepositoriesApi(),
    private val commitsApi: CommitsApi = CommitsApi(),
    private val uuid: String = UUID.randomUUID().toString().replace("-","")
) {
    fun commit(container: String, message: String, tags: List<String>) {
        val metadata = Metadata.load(repositoriesApi.getRepository(container).properties)
        val repoStatus = repositoriesApi.getRepositoryStatus(container)
        val sourceCommit = repoStatus.sourceCommit
        val tagMetadata = mutableMapOf<String, String>()
        for (tag in tags) {
            val (key, value) = if (tag.contains("=")) {
                Pair(tag.substringBefore("="), tag.substringAfter("="))
            } else {
                Pair(tag, "")
            }
            tagMetadata[key] = value
        }
        metadata.user = user
        metadata.email = email
        metadata.message = message
        metadata.tags = tagMetadata
        if (sourceCommit != null) {
            metadata.source = sourceCommit
        }
        val commit = Commit(uuid, metadata.toMap())
        val response = commitsApi.createCommit(repository, commit)
        val hash = response.id
        println("Commit $hash")
    }

}

File: internal/app/providers/common/Commit.go
package common

import (
	"fmt"
	"github.com/google/uuid"
	client "github.com/titan-data/titan-client-go"
	"strconv"
	"strings"
)

func Commit(repo string, message string, tags []string, user string, email string, port int) {
	cfg.BasePath = "http://localhost:" + strconv.Itoa(port)

	repoProps, _, _ := repositoriesApi.GetRepository(ctx, repo)
	metadata := Metadata{}.Load(repoProps.Properties)
	repoStatus, _, _ := repositoriesApi.GetRepositoryStatus(ctx, repo) //TODO handle this error
	sourceCommit := repoStatus.SourceCommit
	tagMetadata := make(map[string]string)
	for _, tag := range tags {
		var k string
		var v string
		if strings.Contains(tag, "=") {
			s := strings.Split(tag, "=")
			k = s[0]
			v = s[1]
		} else {
			k = tag
			v = ""
		}
		tagMetadata[k] = v
	}
	metadata.SetEmail(email)
	metadata.SetUser(user)
	metadata.SetMessage(message)
	metadata.SetTags(tagMetadata)
	metadata.SetSource(sourceCommit)

	guid := uuid.New().String()
	guid = strings.ReplaceAll(guid, "-", "")
	commit := client.Commit{
		Id:         guid,
		Properties: metadata.ToMap(),
	}
	response, _, _ := commitsApi.CreateCommit(ctx, repo, commit)
	fmt.Println("Commit " + response.Id)
}